url9d32402837_dusan_rychnovsky_chess_endgames_tgz-pJ8-cz_dusanrychnovsky_chessendgames_core_strategies_PrecomputedValuesJ8/src/cz/dusanrychnovsky/chessendgames/core/strategies/PrecomputedValues.java

```diff
@@ -0,0 +1,329 @@
+			count = result.walkthrough();
+			long finishTime = System.currentTimeMillis();
+			
+			long duration = finishTime - startTime; 
+			System.out.println(duration + " ms");
+			
+			System.out.println(count + " new records");
+		}
+		while (count != 0);
+		
+		// the cache won't be needed anymore - allow to free the occupied memory
+		result.cache.clear();
+		
+		return result;
+	}
+	
+	/**
+	 * 
+	 * @param whiteKing
+	 * @param blackKing
+	 * @param blackRook
+	 */
+	private PrecomputedValues(King whiteKing, King blackKing, Rook blackRook)
+	{
+		this.whitePlayer = whiteKing.getPlayer();
+		this.whiteKing = whiteKing;
+		
+		this.blackPlayer = blackKing.getPlayer();
+		this.blackKing = blackKing;
+		this.blackRook = blackRook;
+	}
+	
+	/**
+	 * 
+	 * @param file
+	 * @throws IOException
+	 */
+	public void save(File file) throws IOException
+	{
+		ObjectOutputStream out = null;
+		
+		try {
+			
+			out = new ObjectOutputStream(new BufferedOutputStream(new FileOutputStream(file)));
+			save(out);
+		}		
+		finally {
+			if (out != null) {
+				out.close();
+			}
+		}
+	}
+	
+	/**
+	 * 
+	 * @param out
+	 * @throws IOException
+	 */
+	public void save(ObjectOutput out) throws IOException {
+		out.writeObject(this);
+	}
+	
+	/**
+	 * 
+	 * @param file
+	 * @return
+	 * @throws IOException
+	 * @throws ClassNotFoundException
+	 */
+	public static PrecomputedValues load(File file) throws IOException, ClassNotFoundException
+	{
+		ObjectInputStream in = null;
+		
+		try {
+			in = new ObjectInputStream(new BufferedInputStream(new FileInputStream(file)));
+			return load(in);
+		}
+		finally {
+			if (in != null) {
+				in.close();
+			}
+		}
+	}
+	
+	/**
+	 * 
+	 * @param in
+	 * @return
+	 * @throws IOException
+	 * @throws ClassNotFoundException
+	 */
+	public static PrecomputedValues load(ObjectInputStream in) throws IOException, ClassNotFoundException
+	{
+		PrecomputedValues result = (PrecomputedValues) in.readObject();	
+		return result;
+	}
+	
+	/**
+	 * 
+	 * @return
+	 */
+	private int walkthrough()
+	{
+		int result = 0;
+		
+		for (Position whiteKingPos : Position.getAllPositions())
+		{
+			for (Position blackKingPos : Position.getAllPositions())
+			{
+				if (whiteKingPos == blackKingPos) {
+					continue;
+				}
+				
+				for (Position blackRookPos : Position.getAllPositions())
+				{
+					if (blackRookPos == blackKingPos || blackRookPos == whiteKingPos) {
+						continue;
+					}
+					
+					Situation situation = new Situation();
+					situation.addPiece(whiteKing, whiteKingPos);
+					situation.addPiece(blackKing, blackKingPos);
+					situation.addPiece(blackRook, blackRookPos);
+					
+					if (evaluate(situation, blackPlayer)) {
+						result++;
+					}
+					
+					if (evaluate(situation, whitePlayer)) {
+						result++;
+					}
+				}
+			}
+		}
+		
+		return result;
+	}
+	
+	/**
+	 * 
+	 * @param situation
+	 * @param player
+	 * @return
+	 */
+	public boolean evaluate(Situation situation, Player player)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.